lib/Dialect/FIRRTL/Transforms/LowerXMR.cpp

```diff
@@ -25,95 +26,114 @@ using namespace circt;
-            // Get the dataflow value into the src.
-            if (auto remoteOpPath = getRemoteRefSend(connect.getSrc()))
-              reachingRefSendAt[connect.getDest()] = remoteOpPath;
-            else
-              return failure();
-            opsToRemove.push_back(connect);
+            markForRemoval(connect);
+            // Merge the dataflow classes of destination into the source of the
+            // Connect. This handles two cases:
+            // 1. If the dataflow at the source is known, then the
+            // destination is also inferred. By merging the dataflow class of
+            // destination with source, every value reachable from the
+            // destination automatically infers a reaching RefSend.
+            // 2. If dataflow at source is unkown, then just record that both
+            // source and destination will have the same dataflow information.
+            // Later in the pass when the reaching RefSend is inferred at the
+            // leader of the dataflowClass, then we automatically infer the
+            // dataflow at this connect and every value reachable from the
+            // destination.
+            dataFlowClasses.unionSets(
+                dataFlowClasses.getOrInsertLeaderValue(connect.getSrc()),
+                dataFlowClasses.getOrInsertLeaderValue(connect.getDest()));
+            return success();
+          })
+          .Case<RefResolveOp>([&](RefResolveOp resolve) {
+            // Merge dataflow, under the same conditions as above for Connect.
+            // 1. If dataflow at the resolve.getRef is known, propagate that to
+            // the result. This is true for downward scoped XMRs, that is,
+            // RefSendOp must be visited before the corresponding RefResolveOp
+            // is visited.
+            // 2. Else, just record that both result and ref should have the
+            // same reaching RefSend. This condition is true for upward scoped
+            // XMRs. That is, RefResolveOp can be visited before the
+            // corresponding RefSendOp is recorded.
+            dataFlowClasses.unionSets(
+                dataFlowClasses.getOrInsertLeaderValue(resolve.getRef()),
+                dataFlowClasses.getOrInsertLeaderValue(resolve.getResult()));
+            resolveOps.push_back(resolve);
+            markForRemoval(resolve);
             return success();
           })
-          .Case<RefResolveOp>(
-              [&](auto resolve) { return handleRefResolve(resolve); })
           .Default([&](auto) { return success(); });
     };
 
-    DenseSet<Operation *> visitedModules;
     // Traverse the modules in post order.
     for (auto node : llvm::post_order(&instanceGraph)) {
       auto module = dyn_cast<FModuleOp>(*node->getModule());
       if (!module)
         continue;
-      if (!visitedModules.insert(module).second)
-        continue;
-
       LLVM_DEBUG(llvm::dbgs()
                  << "\n Traversing module:" << module.moduleNameAttr());
```

Comment:
Just curious, why the newline at the beginning not the end?

Reply:
It was random, will move it.